packages/core/src/directive/directive.ts

```diff
@@ -20,13 +22,37 @@ export class Directive<T extends keyof DirectiveArgs = any> {
   params(): Param[] {
     if (!this.hasArgs()) return [];
 
+    if (isParamMap(this.args)) {
+      return Object.values(this.args)
+    }
+
     return Array.isArray(this.args)
+        // Assumes args is a array of objects which have a "params" method which returns Param[]
       ? this.args.reduce((r, x) => [...r, ...x.params()], [])
+       // Assumes args has a "params" method which returns Param[]
+       // @ts-expect-error
       : this.args.params();
+
+    // TODO: The assumptions above are not type safe. We need an interface that defines a params() method which returns Param[].
+    // this.args should then be contrained by a type which unites the above and ParamMap.
```

Comment:
@binier  Although I didn't change any of this code (added comments only), the compiler began to complain and threw an error:

![Screen Shot 2020-10-22 at 1 59 53 PM](https://user-images.githubusercontent.com/3927939/96933238-f68b0580-1474-11eb-8777-bf6f3f93c711.png)

Probably due to the fact that `RecurseArgs` is possibly by typedef
```
{
  loop?: boolean 
  depth?: number
}
```

Now, I'm not sure why it needs to support that. If `RecurseArgs` typedef was just 

```
export type RecurseArgs = undefined | {
  loop?: Param<'boolean'>;
  depth?:  Param<'int'>;
}
```

We wouldn't have this problem.

Reply:
Because **loop** and **depth** might be **raw** values instead of necessarily be [**graphql vars**](https://dgraph.io/docs/query-language/graphql-variables). Note that `Param` = `GraphQL Var`.

Reply:
Since this supports existing feature which is it's not entirely clear to me, can you do this?